subsys/usb/class/dfu/usb_dfu.c

```diff
@@ -407,7 +427,20 @@ static int dfu_class_handle_to_host(struct usb_setup_packet *setup,
 			dfu_data.status, dfu_data.state);
 
 		if (dfu_data.state == dfuMANIFEST_SYNC) {
+
+#if IS_ENABLED(CONFIG_USB_DFU_REBOOT)
+			dfu_data.state = dfuMANIFEST_WAIT_RST;
+
+			LOG_DBG("Scheduling reboot in 500ms");
+
+			/*
+			 * Reboot with a delay,
+			 * so there is some time to send the status to the host
+			 */
+			k_work_schedule_for_queue(&USB_WORK_Q, &reboot_work, K_MSEC(500));
+#else
```

Comment:
Maybe it will be a bit clearer that way:

```diff
diff --git a/subsys/usb/class/dfu/usb_dfu.c b/subsys/usb/class/dfu/usb_dfu.c
index 1fe835cd020..42ead13b49e 100644
--- a/subsys/usb/class/dfu/usb_dfu.c
+++ b/subsys/usb/class/dfu/usb_dfu.c
@@ -99,10 +99,6 @@ struct dfu_worker_data_t {
 
 static struct dfu_worker_data_t dfu_data_worker;
 
-#ifdef CONFIG_USB_DFU_REBOOT
-static struct k_work_delayable reboot_work;
-#endif
-
 struct usb_dfu_config {
        struct usb_if_descriptor if0;
        struct dfu_runtime_descriptor dfu_descr;
@@ -414,6 +410,30 @@ static void dfu_timer_expired(struct k_timer *timer)
        }
 }
 
+#ifdef CONFIG_USB_DFU_REBOOT
+static struct k_work_delayable reboot_work;
+
+static void reboot_work_handler(struct k_work *item)
+{
+       ARG_UNUSED(item);
+
+       sys_reboot(SYS_REBOOT_WARM);
+}
+
+static void reboot_schedule(void)
+{
+       dfu_data.state = dfuMANIFEST_WAIT_RST;
+
+       LOG_DBG("Scheduling reboot in 500ms");
+
+       /*
+        * Reboot with a delay,
+        * so there is some time to send the status to the host
+        */
+       k_work_schedule_for_queue(&USB_WORK_Q, &reboot_work, K_MSEC(500));
+}
+#endif
+
 static int dfu_class_handle_to_host(struct usb_setup_packet *setup,
                                    int32_t *data_len, uint8_t **data)
 {
@@ -427,17 +447,8 @@ static int dfu_class_handle_to_host(struct usb_setup_packet *setup,
                        dfu_data.status, dfu_data.state);
 
                if (dfu_data.state == dfuMANIFEST_SYNC) {
-
 #if IS_ENABLED(CONFIG_USB_DFU_REBOOT)
-                       dfu_data.state = dfuMANIFEST_WAIT_RST;
-
-                       LOG_DBG("Scheduling reboot in 500ms");
-
-                       /*
-                        * Reboot with a delay,
-                        * so there is some time to send the status to the host
-                        */
-                       k_work_schedule_for_queue(&USB_WORK_Q, &reboot_work, K_MSEC(500));
+                       reboot_schedule();
 #else
                        dfu_data.state = dfuIDLE;
 #endif
@@ -861,15 +872,6 @@ static void dfu_work_handler(struct k_work *item)
        }
 }
 
-#ifdef CONFIG_USB_DFU_REBOOT
-static void reboot_work_handler(struct k_work *item)
-{
-       ARG_UNUSED(item);
-
-       sys_reboot(SYS_REBOOT_WARM);
-}
-#endif
-
 static int usb_dfu_init(const struct device *dev)
 {
        const struct flash_area *fa;
```